File: multi_agent/projects/im-mobile/lib/services/group_call_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/group_call.dart';

class GroupCallService {
  final String baseUrl;
  final String authToken;

  GroupCallService({required this.baseUrl, required this.authToken});

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $authToken',
        'X-User-Id': authToken,
      };

  Future<GroupCall> createCall(int conversationId, String callType) async {
    final response = await http.post(
      Uri.parse('$baseUrl/calls/group/create'),
      headers: _headers,
      body: jsonEncode({'conversationId': conversationId, 'callType': callType}),
    );
    return GroupCall.fromJson(jsonDecode(response.body));
  }

  Future<void> joinCall(int callId) async {
    await http.post(
      Uri.parse('$baseUrl/calls/group/$callId/join'),
      headers: _headers,
    );
  }

  Future<void> leaveCall(int callId) async {
    await http.post(
      Uri.parse('$baseUrl/calls/group/$callId/leave'),
      headers: _headers,
    );
  }

  Future<void> toggleMute(int callId) async {
    await http.post(
      Uri.parse('$baseUrl/calls/group/$callId/mute'),
      headers: _headers,
    );
  }

  Future<void> toggleVideo(int callId) async {
    await http.post(
      Uri.parse('$baseUrl/calls/group/$callId/video'),
      headers: _headers,
    );
  }

  Future<void> toggleScreenShare(int callId, bool enable) async {
    await http.post(
      Uri.parse('$baseUrl/calls/group/$callId/screen-share'),
      headers: _headers,
      body: jsonEncode({'enable': enable}),
    );
  }

  Future<void> endCall(int callId) async {
    await http.post(
      Uri.parse('$baseUrl/calls/group/$callId/end'),
      headers: _headers,
    );
  }
}

File: im-desktop/src/services/group-call-service.ts
import { GroupCall, GroupCallParticipant, CreateCallRequest } from '../types/group-call';
import { apiClient } from './api-client';

export class GroupCallService {
  async createCall(request: CreateCallRequest): Promise<GroupCall> {
    const response = await apiClient.post('/calls/group/create', request);
    return response.data;
  }

  async joinCall(callId: number): Promise<void> {
    await apiClient.post(`/calls/group/${callId}/join`);
  }

  async leaveCall(callId: number): Promise<void> {
    await apiClient.post(`/calls/group/${callId}/leave`);
  }

  async toggleMute(callId: number): Promise<void> {
    await apiClient.post(`/calls/group/${callId}/mute`);
  }

  async toggleVideo(callId: number): Promise<void> {
    await apiClient.post(`/calls/group/${callId}/video`);
  }

  async toggleScreenShare(callId: number, enable: boolean): Promise<void> {
    await apiClient.post(`/calls/group/${callId}/screen-share`, { enable });
  }

  async getCallInfo(callId: number): Promise<{ call: GroupCall; participants: GroupCallParticipant[] }> {
    const response = await apiClient.get(`/calls/group/${callId}`);
    return response.data;
  }

  async endCall(callId: number): Promise<void> {
    await apiClient.post(`/calls/group/${callId}/end`);
  }
}

export const groupCallService = new GroupCallService();
